### PR TITLE
[Customer][Shop] See all addresses in address book 

### PR DESCRIPTION
**Pull request by tuka217**
| Q               | A
| --------------- | ---
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| Related tickets | -
| License         | MIT

### CODE REVIEW

File: features/account/customer_account/address_book/viewing_addresses.feature [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+
+    Background:
+        Given the store operates on a single channel in "United States"
+        And I am a logged in customer
+        And I have an address "Lucifer Morningstar", "Seaside Fwy", "90802", "Los Angeles", "United States" in address book
+
+    @ui
+    Scenario: Viewing all addresses
+        When I browse my addresses
+        Then I should see a single address in a list
```
Comment (pjedrzejewski) [inline features/account/customer_account/address_book/viewing_addresses.feature]: in the list

File: src/Sylius/Bundle/CoreBundle/Doctrine/ORM/AddressRepository.php [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+ */
+class AddressRepository extends EntityRepository implements AddressRepositoryInterface
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function findByCustomer(CustomerInterface $customer)
+    {
+        return $this->createQueryBuilder('o')
+            ->leftJoin('o.customers','customer')
```
Comment (pjedrzejewski) [inline src/Sylius/Bundle/CoreBundle/Doctrine/ORM/AddressRepository.php]: customers? I am pretty sure we have ``o.customer`` relation only!

File: features/account/customer_account/address_book/viewing_addresses.feature [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+    As a Customer
+    I want to be able to browse my addresses
+
+    Background:
+        Given the store operates on a single channel in "United States"
+        And I am a logged in customer
+        And I have an address "Lucifer Morningstar", "Seaside Fwy", "90802", "Los Angeles", "United States" in address book
+
+    @ui
+    Scenario: Viewing all addresses
```
Comment (pjedrzejewski) [inline features/account/customer_account/address_book/viewing_addresses.feature]: Can we also add another customer in the background with some addresses in the address book? In order to make sure that they see only their addresses. (that's +1 scenario at least)

File: src/Sylius/Behat/Context/Ui/Shop/AccountContext.php [hunk 4]
```diff
@@ -459,4 +487,5 @@ private function assertFieldValidationMessage(PageInterface $page, $element, $ex
             sprintf('There should be a message: "%s".', $expectedMessage)
         );
     }
+
```
Comment (Zales0123) [inline src/Sylius/Behat/Context/Ui/Shop/AccountContext.php]: Why? :(

File: src/Sylius/Behat/Context/Setup/CustomerContext.php [hunk 5]
```diff
@@ -176,6 +176,27 @@ public function heHasBillingAddress(CustomerInterface $customer, AddressInterfac
...
+     */
+    public function theCustomerSubscribedToTheNewsletter(CustomerInterface $customer)
+    {
+        $customer->setSubscribedToNewsletter(true);
+
+        $this->customerManager->flush();
+    }
+
+    /**
+     * @Given /^I have an (address "([^"]+)", "([^"]+)", "([^"]+)", "([^"]+)", "([^"]+)") in address book$/
```
Comment (pamil) [inline src/Sylius/Behat/Context/Setup/CustomerContext.php]: `(address "([^"]+)", "([^"]+)", "([^"]+)", "([^"]+)", "([^"]+)")` -> `(address "[^"]+", "[^"]+", "[^"]+", "[^"]+", "[^"]+")`

File: features/account/customer_account/address_book/viewing_addresses.feature (hunk 3 above)
Comment (tuka217) [inline features/account/customer_account/address_book/viewing_addresses.feature]: Ok

File: src/Sylius/Bundle/ShopBundle/Resources/translations/messages.en.yml [hunk 6]
```diff
@@ -10,3 +10,4 @@ sylius:
                 personal_information: Personal information
                 change_password: Change password
                 order_history: Order history
+                address_book: Address book
```
Comment (Zales0123) [inline src/Sylius/Bundle/ShopBundle/Resources/translations/messages.en.yml]: If you're changing this file, maybe consider sorting this messages alphabetically? :)

File: src/Sylius/Behat/Context/Ui/Shop/AccountContext.php (hunk 4 above)
Comment (pamil) [inline src/Sylius/Behat/Context/Ui/Shop/AccountContext.php]: Why?

File: src/Sylius/Bundle/CoreBundle/Doctrine/ORM/AddressRepository.php [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+            ->where('customer.id = customer_id')
```
Comment (pamil) [inline src/Sylius/Bundle/CoreBundle/Doctrine/ORM/AddressRepository.php]: `customer_id` -> `:customer_id`

File: src/Sylius/Bundle/ShopBundle/Resources/views/Account/AddressBook/index.html.twig [hunk 8]
```diff
@@ -0,0 +1,33 @@
+{% extends '@SyliusShop/Account/layout.html.twig' %}
+{% import '@SyliusUi/Macro/messages.html.twig' as messages %}
```
Comment (Zales0123) [inline src/Sylius/Bundle/ShopBundle/Resources/views/Account/AddressBook/index.html.twig]: I would add blank line before.

File: src/Sylius/Bundle/ShopBundle/Resources/views/Account/AddressBook/index.html.twig [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+        <h1 class="ui dividing header">
+            {{ 'sylius.ui.address_book'|trans }}
+            <div class="sub header">{{ 'sylius.ui.browse_your_addresses'|trans }}</div>
+        </h1>
+        <div class="ui stackable grid">
+            <div class="ten wide column">
+                {{ messages.info('sylius.ui.no_default_address'|trans) }}
+            </div>
+            <div class="six wide column">
+                {% if(addresses is empty) %}
```
Comment (Zales0123) [inline src/Sylius/Bundle/ShopBundle/Resources/views/Account/AddressBook/index.html.twig]: We don't need any parenthesis here, I suppose.

File: src/Sylius/Bundle/UiBundle/Resources/translations/messages.en.yml [hunk 10]
```diff
@@ -465,8 +467,10 @@ sylius:
         no_label: 'No'
         no_account: 'No account yet'
         no_account: no account
+        no_addresses: 'There is no addresses to display'
```
Comment (NeverResponse) [inline src/Sylius/Bundle/UiBundle/Resources/translations/messages.en.yml]: either `are ... addresses` or `is ... address`

File: src/Sylius/Behat/Context/Transform/CustomerContext.php [hunk 11]
```diff
@@ -72,6 +72,7 @@ public function getOrCreateCustomerByEmail($email)
 
     /**
      * @Transform /^(he|his|she|her|the customer of my account)$/
+     * @Transform /^(this customer)$/
```
Comment (NeverResponse) [inline src/Sylius/Behat/Context/Transform/CustomerContext.php]: Shouldn't the `SharedStorageContext` take care of this?

File: features/account/customer_account/address_book/viewing_addresses.feature [hunk 12]
```diff
@@ -0,0 +1,23 @@
+@customer_account
+Feature: Viewing customer's addresses
+    In order to see all my addresses in address book
+    As a Customer
+    I want to be able to browse my addresses
+
+    Background:
+        Given the store operates on a single channel in "United States"
+        And there is a customer "John Doe" identified by an email "doe@example.com" and a password "banana"
+        And this customer has an address "John Doe", "Banana Street", "90232", "New York", "United States" in address book
```
Comment (Zales0123) [inline features/account/customer_account/address_book/viewing_addresses.feature]: Are this two steps required for this feature?

File: src/Sylius/Behat/Context/Setup/CustomerContext.php [hunk 13]
```diff
@@ -176,6 +176,37 @@ public function heHasBillingAddress(CustomerInterface $customer, AddressInterfac
...
+
+    /**
+     * @Given /^I have an (address "[^"]+", "[^"]+", "[^"]+", "[^"]+", "[^"]+") in address book$/
+     */
+    public function iHaveAnAddressInAddressBook(AddressInterface $address)
+    {
+        $user = $this->sharedStorage->get('user');
+        $user->getCustomer()->addAddress($address);
+        
+        $this->customerManager->flush();
```
Comment (Zales0123) [inline src/Sylius/Behat/Context/Setup/CustomerContext.php]: What say you for:
```php
$user = $this->sharedStorage->get('user');
$this->thisCustomerHasAnAddressInAddressBook($user->getCustomer(), $address);
```
?

File: src/Sylius/Behat/Page/Shop/Account/AddressBook/IndexPage.php [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+    {
+        return null !== $this->getElement('addresses')->find('css', sprintf('.title:contains("%s")', $fullName));
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function hasNoExistingAddressesMessage()
+    {
+        return 'Info There are no addresses to display' === $this->getDocument()->find('css', '#addresses > .message')->getText();
```
Comment (Zales0123) [inline src/Sylius/Behat/Page/Shop/Account/AddressBook/IndexPage.php]: A little to long line, imo.

File: features/account/customer_account/address_book/viewing_addresses.feature (hunk 12 above)
Comment (tuka217) [inline features/account/customer_account/address_book/viewing_addresses.feature]: Yups, beacause we want to be sure, that you will see only you addresess on the list. In scenario about viewing addresses I can add an step "And I should not see address assigned to "John Doe".

Comment (Zales0123) [inline features/account/customer_account/address_book/viewing_addresses.feature]: That would be better 👍  As it was pretty confused for me, why we need info about some John Doe ;)

**Comment by michalmarcinkowski** [comment]
Closing in favour of #6472

**Pull request closed without merging**